### DISCUSSION TITLE
removeDuplicateRules

### DISCUSSION THREAD

**Discussion by Uvatek** [jakubpawlowicz/clean-css#1188] [Q&A] [answered]
Hello community!

I discovered this beautiful project but encountered a problem (maybe out of ignorance)
I do a test with 2 sames CSS files to merge and clean duplicates.
test.css : a{font-weight:bold;}
test2.css : a{font-weight:bold;}

`var options =  { 
level: {
    1: {
      removeEmpty: true
    },
    2: {
      all: false,
      removeDuplicateRules: true,
      removeUnusedAtRules: true
    }
  }
};
var output = new CleanCSS(options).minify(['test.css', 'test2.css'])`

I get this:
`a{}a{font-weight:700}`

Instead of that :
`a{font-weight:700}`

To get the good result I can call a second time with removeEmpty:true but it seems to me to be a walk around..

Did i make a mistake?

**Comment by jakubpawlowicz** [comment, marked as answer]
Hello @Uvatek !

There are two `removeEmpty` flags - in level 1 and level 2. In your case you can skip level 1 flag and just use level 2 flag:

```
var options = { level: { 2: { all: false, removeEmpty: true, removeDuplicateRules: true, removeUnusedAtRules: true } } }; var output = new CleanCSS(options).minify(['test.css', 'test2.css'])
```

The reason is only level 2 optimizations can figure out `a{font-weight:700}` is repeated - in fact it's `font-weight: 700` which is overridden in second `a{...` block so it can be dropped in first one, then in another pass the first `a{}` can be dropped.

Hope this helps!

**Reply by Uvatek** [reply]
I tried by putting everything in the property level 1 it's good also..
Well if I understood correctly the properties "1" and "2" of the documentation are in reality iterations of treatments and all the properties which are described there can be used each time.. 

And your example works perfectly thank you so much @jakubpawlowicz !

**Reply by jakubpawlowicz** [reply]
Glad it works!

The properties `1` and `2` are for different optimization levels. When you moved all in `2` to `1` then the default in `2` is to do almost all optimizations, so that's why you got the same result.

In general level 1 optimizations operate within a CSS rule, so for example turning `font-weight:bold` to `font-weight:700`, while level 2 optimizations can merge, reorder, and/or drop whole rules.

See https://github.com/clean-css/clean-css#what-level-2-optimizations-do for more details on level 2 optimizations.
